client/src/app/auth/auth.service.ts

```diff
@@ -1,66 +1,183 @@
-import { Injectable } from '@angular/core';
-import { AUTH_CONFIG } from './auth0-variables';
-import * as auth0 from 'auth0-js';
+import {Injectable} from '@angular/core';
+import {AUTH_CONFIG} from './auth0-variables';
 import {Router} from "@angular/router";
+import {HttpClient, HttpHeaders} from "@angular/common/http";
+import {Observable, of, Subscription} from "rxjs";
+import {AuthConfig, OpenIdConfig, UserInfo} from "./auth.model";
+import {mergeMap} from "rxjs/operators";
+import {TimerObservable} from "rxjs-compat/observable/TimerObservable";
+
 
 @Injectable()
 export class AuthService {
 
-  auth0 = new auth0.WebAuth({
-    clientID: AUTH_CONFIG.clientID,
-    domain: AUTH_CONFIG.auth0,
-    responseType: 'token id_token',
-    audience: AUTH_CONFIG.apiUrl,
-    redirectUri: AUTH_CONFIG.callbackURL,
-    scope: 'openid profile read:profile email'
-  });
-
-  userProfile: any;
+  config: AuthConfig;
+  openIdConfig: OpenIdConfig;
+  silentAuthInterval: number;
+  private silentAuthTimer: Observable<number>;
+  private readonly tokenDuration: number
+  private readonly tokenRefreshPeriod: number;
+  private silentAuthSubscription: Subscription;
+  private userInfo$: Observable<UserInfo>;
 
-  constructor(public router: Router) {}
+  constructor(private router: Router, private http: HttpClient) {
+    this.config = AUTH_CONFIG;
+    this.tokenDuration = 900000; // 15 min
+    this.tokenRefreshPeriod = 60000; // 1min
+    this.silentAuthInterval = this.tokenDuration - this.tokenRefreshPeriod;
+  }
 
-  public login(): void {
-    if(this.isAuthenticated()){
-      alert('You are already logged in. Redirecting to homepage...')
-      this.router.navigate(['/home']);
+  getOpenIdConfig(): Observable<OpenIdConfig> {
+    if (this.openIdConfig) {
+      return of(this.openIdConfig);
     } else {
-      this.authorize()
+      return this.http.get(`https://${this.config.domain}/.well-known/openid-configuration`)
+        .map((response) => response as OpenIdConfig);
     }
   }
 
-  public authorize():void{
-    window.location.href = `https://${AUTH_CONFIG.domain}/oauth/authorize?redirect_uri=${AUTH_CONFIG.callbackURL}`;
+  getUserInfo(): Observable<UserInfo> {
+    if (this.userInfo$){
+      return this.userInfo$;
+    }
+    return this.getOpenIdConfig().pipe(mergeMap((openIdConfig) => {
```

Comment:
Would `map` work here as well, instead of `mergeMap`?

Reply:
Actually, i think i should concatMap here because I want the getOpenIdConfig observable to emit a value first before I get the user info. I think map would work but the higher order mapping operators (e.g. mergeMap, concatMap, switchMap etc) is recommended if dealing with multiple observables. https://blog.angular-university.io/rxjs-higher-order-mapping/